{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# File Name: Jones_Shweta_randomizedMotifSearch.ipynb\n",
    "#### Description: \n",
    "The program is given a file if fasta sequences, and calculates the consensus of the file and its relative entropy score. \n",
    "#### Sample CommandLine Entry: \n",
    "python Jones_Shweta_randomizedMotifSearch.py -i int -k int -p float <input.fa>output.out\n",
    "##### stdin: \n",
    "fasta sequence containing DNA sequences\n",
    "##### stdout: \n",
    "the relative entropy score and consensus\n",
    "#### Author: \n",
    "Shweta Jones\n",
    "#### Group: \n",
    "Sahasra Shankar, Kritin Nandish, Maheep Luthra\n",
    "#### History:      \n",
    "sj 08/08/2021 Created\n",
    "sj 08/09/2021 Updated\n",
    "sj 08/10/2021 Updated"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Class Descriptions \n",
    "#### FastAReader\n",
    "This class reads in the FastA file and provides the main the header and sequence in the FastA file. \n",
    "#### CommandLine\n",
    "This class deals with the commandLine options provided and sets the values to be used by main and other classes. \n",
    "#### Calculations\n",
    "The calculations class deals with all the processing related to the sequences within the FastA file. The class creates the null profile, finds random motifs, creates a profile, iterate through motifs, and caclualtes the relative entropy score. \n",
    "#### Main\n",
    "This class runs the entire program by reading the command line and outputting the corresponding profile and relative entropy score. The input is the options given by the user through the commandLine. The main function prints out the consensus with the highest relative entropy score and its score into the specified output file. \n",
    " "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AGAACTGT   3.448259051086194\n"
     ]
    }
   ],
   "source": [
    "# #########################################################################\n",
    "# File: Jones_Shweta_randomizedMotifSearch.ipynb\n",
    "#  executable: randomizedMotifSearch.py\n",
    "# Purpose: To determine the relative entropy score and corresponding consensus given a fasta sequence\n",
    "#   stdin: fasta sequence containing DNA sequences\n",
    "#   stderr: errors and status\n",
    "#   stdout: the relative entropy score and consensus\n",
    "# Author: Shweta Jones\n",
    "# Group: Sahasra Shankar, Kritin Nandish, Maheep Luthra\n",
    "# History:      sj 08/08/2021 Created\n",
    "#               sj 08/09/2021 Updated\n",
    "#               sj 08/10/2021 Updated\n",
    "# #########################################################################\n",
    "import sys\n",
    "from random import random\n",
    "import random\n",
    "import math\n",
    "class FastAreader:\n",
    "    ''' This class reads in the FastA file and provides the main the header and sequence in the FastA file'''\n",
    "    def __init__(self, fname=''):\n",
    "        '''contructor: saves attribute fname '''\n",
    "        self.fname = fname\n",
    "    def doOpen(self):\n",
    "        ''' Opens the file in stdin '''\n",
    "        if self.fname is '':\n",
    "            return sys.stdin\n",
    "        else:\n",
    "            return open(self.fname)\n",
    "    def readFasta(self):\n",
    "        ''' Reads the file and returns the header and sequence one by one '''\n",
    "        header = ''\n",
    "        sequence = ''\n",
    "        with self.doOpen() as fileH:\n",
    "            header = ''\n",
    "            sequence = ''\n",
    "            # skip to first fasta header\n",
    "            line = fileH.readline()\n",
    "            while not line.startswith('>'):\n",
    "                line = fileH.readline()\n",
    "            header = line[1:].rstrip()\n",
    "            for line in fileH:\n",
    "                if line.startswith('>'):\n",
    "                    yield header, sequence\n",
    "                    header = line[1:].rstrip()\n",
    "                    sequence = ''\n",
    "                else:\n",
    "                    sequence += ''.join(line.rstrip().split()).upper()\n",
    "        yield header, sequence\n",
    "\n",
    "class CommandLine():\n",
    "    ''' Deals with the commandLine options provided and sets the values to be used by main and other classes '''\n",
    "    def __init__(self, inOpts=None):\n",
    "        ''' Contains all the information that either saves the information given by the user, or uses the default values'''\n",
    "        import argparse\n",
    "        self.parser = argparse.ArgumentParser(description='This program find the motif with the highest relative entropy score', add_help=True)\n",
    "        self.parser.add_argument('-i', '--iterations', type=int, action='store', default=1000, dest='iterations',\n",
    "                                 help='Provide the number of runs the motif search should take, the default is 1000 iterations.') #handles the iterations option provided by the user, or set as the default\n",
    "        self.parser.add_argument('-p', '--pseudoCount', type=float, action='store', default=1,dest='pseudoCount',\n",
    "                                 help='Provide the psuedo count, the default is 1 pseudo count.') #handles the pseudo count option provided by the user, or set as the default\n",
    "        self.parser.add_argument('-k', '--kmerLength', type=int, action='store', default=13, dest='kmerLength',\n",
    "                                 help='Provide the kmer length') #handles the kmer length option provided by the user, or set as the default\n",
    "        if inOpts is None:\n",
    "            self.args = self.parser.parse_args()\n",
    "        else :\n",
    "            self.args = self.parser.parse_args(inOpts)\n",
    "\n",
    "class Calculations:\n",
    "    ''' The calculations class deals with all the processing related to the sequences '''\n",
    "    def __init__(self):\n",
    "        ''' Description: Initializes all the values that will be utilized within the class '''\n",
    "        ''' Input: The method does not have any inputs '''\n",
    "        ''' Output: The method initializes certain values that will be changed in future methods '''\n",
    "        self.aCount = 0; self.tCount = 0; self.cCount = 0; self.gCount = 0 # Initializes the counts of all the nucleotides\n",
    "        self.aNull = 0; self.tNull = 0; self.cNull = 0; self.gNull = 0 # Initializes the null values of all the nucleotides\n",
    "        self.total = 0 # Initializes the total count of all nucleotides \n",
    "        self.profArray = [] # Initializes the array of profiles\n",
    "\n",
    "    def createNull(self, a, t, c, g, pCount):\n",
    "        ''' Description: Calculates the Null Model '''\n",
    "        ''' Input: This method uses the number of A, T, C and G and the pseudo count '''\n",
    "        ''' Output: The method saves the null values of each of the nucleotides to be used in other methods '''\n",
    "        self.aCount += a; self.tCount += t; self.cCount += c; self.gCount += g # Saves the nucleotide count to the initializes self value\n",
    "        self.total = a + t + c + g + (pCount*4) # Determines the total number of nucleotides with psuedocounts\n",
    "        self.aNull = ((a+pCount)/self.total); self.tNull = ((t+pCount)/self.total); self.cNull = ((c+pCount)/self.total); self.gNull = ((g+pCount)/self.total) # Calculates the null probabilities of each of the nucleotides\n",
    "\n",
    "    def randomMotif(self, seqList, kmerLength):\n",
    "        ''' Description: Retrieves random motifs of a certain length in each of the sequences '''\n",
    "        ''' Input: This method uses a list of sequences and the length of the k-mer which is specified by the user '''\n",
    "        ''' Output: The method returns a random motif within each of the sequences '''\n",
    "        totalLength = len(seqList[0])\n",
    "        ranList = [random.randint(0, totalLength-(kmerLength+1)) for i in seqList] # Creates a list of random motifs that could be anywhere from the beginning of the sequence to the end-kmerlength of the sequence \n",
    "        motifList = []; count = 0\n",
    "        for i in seqList:\n",
    "            motifList.append(i[ranList[count]:ranList[count]+kmerLength]) # Finds and saves the selected motif into a list of motifs\n",
    "            count+=1\n",
    "        return motifList\n",
    "\n",
    "    def createProfile(self, motifList, kmerLength, pseudoCount):\n",
    "        ''' Description: This method determines the profile given the list of motifs and length of the k-mer '''\n",
    "        ''' Input: The input for this method is the list of motif, the length of the k-mer, and the pseudo count '''\n",
    "        ''' Output: The method returns the profile '''\n",
    "        self.profArray = []; subList = []\n",
    "        for i in range(kmerLength): # Creates an empty list of lists\n",
    "            subList.append([])\n",
    "        numMotif = len(motifList)\n",
    "        aProfList = []; tProfList = []; cProfList = []; gProfList = [];\n",
    "        for i in range(kmerLength): # Creates a 2D array to create the profile by counting the number of nucleotides in specific positions\n",
    "            for j in range(numMotif):\n",
    "                subList[i].append(motifList[j][i]) \n",
    "            aProfList.append((subList[i].count(\"A\")+pseudoCount)/(numMotif+(pseudoCount*4)))\n",
    "            tProfList.append((subList[i].count(\"T\")+pseudoCount)/(numMotif+(pseudoCount*4)))\n",
    "            cProfList.append((subList[i].count(\"C\")+pseudoCount)/(numMotif+(pseudoCount*4)))\n",
    "            gProfList.append((subList[i].count(\"G\")+pseudoCount)/(numMotif+(pseudoCount*4)))\n",
    "        self.profArray.append(aProfList); self.profArray.append(tProfList); self.profArray.append(cProfList); self.profArray.append(gProfList) # Places the probabilities into a profile array\n",
    "        return self.profArray\n",
    "\n",
    "    def iterMotifs(self, seqList, kmerLength):\n",
    "        ''' Description: Determines all the motifs in a given set of sequences, and determines a motif with the highest score given the profile '''\n",
    "        ''' Input: The method uses the all the sequences and the length of the k-mer '''\n",
    "        ''' Output: The output is the final motif that has the highest value '''\n",
    "        totalLength = len(seqList[0])\n",
    "        finalMotif = []\n",
    "        for i in range(len(seqList)): # Iterates through all the sequences to find the best possible motif for each sequence\n",
    "            motifList = []\n",
    "            for j in range(totalLength-kmerLength): # Finds all the possible motifs in a given sequence\n",
    "                motifList.append(seqList[i][j:j+kmerLength])\n",
    "            motifDict = {}\n",
    "            for k in range(len(motifList)): # Iterates through all the motifs and calculates based on the profile\n",
    "                count = 1\n",
    "                for l in range(kmerLength): # Iterates through the length of the kmer and calculates it's score based on profile\n",
    "                    if (motifList[k][l]) == \"A\": count *=(self.profArray[0][l])\n",
    "                    elif (motifList[k][l]) == \"T\": count *= (self.profArray[1][l])\n",
    "                    elif (motifList[k][l]) == \"C\": count *= (self.profArray[2][l])\n",
    "                    elif (motifList[k][l]) == \"G\": count *= (self.profArray[3][l])\n",
    "                motifDict[motifList[k]] = count # Places the motif and it's correpsonding score in a dictionary\n",
    "            finalMotif.append(max(motifDict, key=motifDict.get)) # Determines the motif with the highest value and adds it to a list of motifs\n",
    "        return finalMotif\n",
    "\n",
    "    def relativeEntropy(self, profile, consensus):\n",
    "        ''' Description: Calculates the relative entropy of the consensus sequence '''\n",
    "        ''' Input: The function uses the profile of the final motif and the determined consensus '''\n",
    "        ''' Output: The method returns the relative entropy score of the consensus '''\n",
    "        score = 0\n",
    "        for i in range(len(consensus)): # Goes through all the nucleotides within the consensus and calculates the total relative entropy score\n",
    "            if consensus[i] == \"A\":\n",
    "                score += (profile[0][i]) * (math.log2((profile[0][i])/self.aNull))\n",
    "            elif consensus[i] == \"T\":\n",
    "                score += (profile[1][i]) * (math.log2((profile[1][i])/self.tNull))\n",
    "            elif consensus[i] == \"C\":\n",
    "                score += (profile[2][i]) * (math.log2((profile[2][i])/self.cNull))\n",
    "            elif consensus[i] == \"G\":\n",
    "                score += (profile[3][i]) * (math.log2((profile[3][i])/self.gNull))\n",
    "        return (score)\n",
    "\n",
    "def main (inFile, myCommandLine=None):\n",
    "    ''' Description: Runs the entire program by reading the command line and outputting the corresponding profile and relative entropy score '''\n",
    "    ''' Input: The input is the options given by the user through the commandLine '''\n",
    "    ''' Output: The main function prints out the consensus with the highest relative entropy score and its score into the specified output file '''\n",
    "    if myCommandLine is None:\n",
    "        myCommandLine = CommandLine()  # read options from the command line\n",
    "    else:\n",
    "        myCommandLine = CommandLine(\n",
    "            myCommandLine)  # interpret the list passed from the caller of main as the commandline.\n",
    "    \n",
    "    seqList = []\n",
    "    aCount = 0; tCount = 0; cCount = 0; gCount = 0\n",
    "    for head, seq in FastAreader(inFile).readFasta(): #iterates through all the sequences within the Fasta\n",
    "        processedSeq = Calculations()\n",
    "        seqList.append(seq)\n",
    "        aCount += seq.count(\"A\"); tCount += seq.count(\"T\"); cCount += seq.count(\"C\"); gCount += seq.count(\"G\")\n",
    "    processedSeq.createNull(aCount, tCount, cCount, gCount, myCommandLine.args.pseudoCount) # Determines the Null Model given the counts of nucleotides within the sequences\n",
    "    allEntropy = []; allConsensus = []\n",
    "    for i in range(myCommandLine.args.iterations): # Iterates through the number of user provided user iterations and outputs the relative entropy score along with its corresponding consensus\n",
    "        oldMotifList = processedSeq.randomMotif(seqList, myCommandLine.args.kmerLength) # Creates the initial motif list using the random motif generator\n",
    "        processedSeq.createProfile(oldMotifList, myCommandLine.args.kmerLength, myCommandLine.args.pseudoCount)\n",
    "        newMotifList = processedSeq.iterMotifs(seqList, myCommandLine.args.kmerLength)\n",
    "        while oldMotifList != newMotifList: # Determines the best motif list which is attained with the new and old motif list are the same\n",
    "            oldMotifList = newMotifList # Saves the last motif list to the oldMotifList \n",
    "            profile = processedSeq.createProfile(oldMotifList, myCommandLine.args.kmerLength, myCommandLine.args.pseudoCount) # Finds the profile of the last motif list\n",
    "            newMotifList = processedSeq.iterMotifs(seqList, myCommandLine.args.kmerLength) # Saves the motif list to a newMotifList to compare\n",
    "        consensus = []\n",
    "        for i in range(myCommandLine.args.kmerLength): # Iterates through the length of the kmer and determines the consensus sequence\n",
    "            aVal = profile[0][i]; tVal = profile[1][i]; cVal = profile[2][i]; gVal = profile[3][i] # Places the number of nucleotides to a variable \n",
    "            maxNuc = ([aVal, tVal, cVal, gVal]).index(max([aVal, tVal, cVal, gVal])) # Determines which nucleotide was in the highest frequence\n",
    "            if maxNuc == 0:nucVal = \"A\"\n",
    "            elif maxNuc == 1:nucVal = \"T\"\n",
    "            elif maxNuc == 2:nucVal = \"C\"\n",
    "            elif maxNuc == 3:nucVal = \"G\"\n",
    "            consensus.append(nucVal) # Appends the nucleotide to the consensus\n",
    "        entropyScore = processedSeq.relativeEntropy(profile, consensus) # Determines the relative entropy score with the profile and consensus\n",
    "        allConsensus.append(consensus) # Appends the consenses to a list of consensus's \n",
    "        allEntropy.append(entropyScore) # Appends the entropy score to a list of entropy's\n",
    "    finalEntropy = max(allEntropy) # Finds the highest entrpy score\n",
    "    finalConsensus = \"\".join(allConsensus[allEntropy.index(finalEntropy)]) # Consolidated the nucleotides within the consenses into one string\n",
    "    print(str(finalConsensus) + \"   \" + str(finalEntropy)) # Prints the consensus and it's entropy score to the output file\n",
    "\n",
    "if __name__== \"__main__\":\n",
    "    main(\"pogCrisprs\", [\"-i 1000\", \"-p 3\", \"-k 8\"])\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Inspection:\n",
    "#### Inspector 1: \n",
    "        You can shorten the length of certain long lines of code, however it looks cleaner the way it is done here\n",
    "#### Inspector 2:\n",
    "        For longer loops I would recommend providing a short description to help understand\n",
    "        Great job on the detailed notes!\n",
    "#### Inspector 3:\n",
    "        The code you have right now looks pretty good but I think there are a few small improvements you can make for better code readability\n",
    "        I noticed that you have docstrings essentially everywhere but I think it would be better to limit docstrings to right under classes and methods and then use comments to specify certain lines of code where what it is doing may not be clear\n",
    "        Maybe also add a few empty lines in between parts of your code so that it is a bit easier to read and figure out what part of your code does what\n",
    "        I would also recommend condensing your docstrings so that the input and output are implied rather then state it explicitly\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
